File: src/annotated_game_of_life.hs
import qualified Data.Map as M

type Coords = (Int, Int)
type Cell = (Bool, Int)
type World = M.Map Coords Cell

tick :: Cell -> Bool
tick (a, 2) = a
tick (_, 3) = True
tick (_, _) = False

explode :: Coords -> World
explode (x, y) =
  M.fromList $
    do xmod <- [-1, 0, 1]
       ymod <- [-1, 0, 1]
       let center = xmod  == 0 && ymod == 0
       return ((x + xmod, y + ymod),
               (center, if center then 0 else 1))

mergeCell :: Cell -> Cell -> Cell
mergeCell (a1, n1) (a2, n2) = (a1 || a2, n1 + n2)

mergeWorlds :: [World] -> World
mergeWorlds = foldr (M.unionWith mergeCell) M.empty

worldTick :: [Coords] -> [Coords]
worldTick coords =
  M.keys $ M.filter tick $ mergeWorlds $ map explode coords

main = print (worldTick [(2,1), (2,2), (2,3)])

-- 6 whitespace
-- 23 loc

File: src/annotated_game_of_life.erl
-module(annotated_game_of_life).
-export([world_tick/1]).

-type coord() :: {integer(), integer()}.
-type cell() :: {boolean(), integer()}.
-type world() :: [[{coord(),cell()}]].

-spec tick(cell()) -> boolean().
tick({A,2}) -> A;
tick({_,3}) -> true;
tick({_,_}) -> false.

-spec explode([coord()]) -> world().
explode(Coords) ->
    [[{{X+DeltaX, Y+DeltaY}, {Center, if Center -> 0; true -> 1 end}}
      || DeltaX <- [-1,0,1],
         DeltaY <- [-1,0,1],
         Center <- [DeltaX =:= 0 andalso DeltaY =:= 0]]
     || {X,Y} <- Coords].

-spec cell_merge(Key::_, cell(), cell()) -> cell().
cell_merge(_K, {A1,N1}, {A2,N2}) -> {A1 or A2, N1+N2}.

-spec world_tick([coord()]) -> [coord()].
world_tick(Coords) ->
    Dict = lists:foldl(fun(D1,D2) -> dict:merge(fun cell_merge/3, D1, D2) end,
                       dict:new(),
                       [dict:from_list(L) || L <- explode(Coords)]),
    dict:fetch_keys(dict:filter(fun(_,V) -> tick(V) end, Dict)).

%% 5 whitespace
%% 24 loc
